### PR TITLE
Generate other functions

### PR DESCRIPTION
**Pull request by EPashkin**
WIP

### CODE REVIEW

**Comment by EPashkin** [comment]
Done

**Comment by EPashkin** [comment]
IMHO `cairo` now will be not empty, I try it.
And next I try features.

File: src/parser.rs [hunk 1]
```diff
@@ -496,7 +496,11 @@ impl Library {
         if func.c_identifier.is_none() {
             return Err(mk_error!("Missing c:identifier attribute", &pos));
         }
-        self.add_function(ns_id, func);
+        let already_added = self.namespace(ns_id).classes_function_cidentifiers
+            .contains(func.c_identifier.as_ref().unwrap());
+        if !already_added {
```
Comment (gkoz) [inline src/parser.rs]: There are duplicate function definitions in girs?

**Comment by EPashkin** [comment]
In `Gio-2.0.gir` two `g_icon_new_for_string` and 30 others

**Comment by gkoz** [comment]
Do all of them have the `moved-to` attribute? We could probably skip those in the codegen.

**Comment by EPashkin** [comment]
Oh missed that, then it can just skipped in simply way. I check

**Comment by EPashkin** [comment]
Can i change result of `Parser::read_function` to `Result<Option<Function>, Error>` or there better way ignore function?

**Comment by gkoz** [comment]
I think we can check the `moved-to` attribute upstream (e.g. in `read_class`) and instead of `read_function` call `ignore_element` then.

**Comment by gkoz** [comment]
To avoid duplication this could be done in a wrapper e.g. `read_function_if_not_moved`

**Comment by EPashkin** [comment]
IMHO we needed this check only for global functions, so i just check `moved-to` beforehand as you say

**Comment by gkoz** [comment]
Makes sense.

**Comment by EPashkin** [comment]
Done
If you think that we don't need feature, we can call GuillaumeGomez

File: src/codegen/sys/ffi_type.rs [hunk 2]
```diff
@@ -72,6 +72,7 @@ fn ffi_inner(env: &Env, tid: library::TypeId, inner: String) -> Result {
                 Type => "GType",
                 Pointer => match &inner[..]  {
                     "void" => "c_void",
+                    "FILE" => return Err(inner),
```
Comment (gkoz) [inline src/codegen/sys/ffi_type.rs]: Looks like `libc` provides `FILE` so we could add it to imports instead.

**Comment by gkoz** [comment]
This still doesn't generate functions on records (e.g.`Array.append_vals` in `GLib`), does it?

**Comment by gkoz** [comment]
About the features, I'll open a PR against rust-gnome to remove the redundant ones.

**Comment by EPashkin** [comment]
Yes, forgot about `g_array_append_vals` and others

**Comment by EPashkin** [comment]
Fixed FILE
About features: we easy can add it to functions.

**Comment by EPashkin** [comment]
I started do record functions

**Comment by EPashkin** [comment]
IMHO Ready for merge

**Comment by gkoz** [comment]
Thanks.
@homu r+

**Pull request merged**